src/data/map.rs

```diff
@@ -0,0 +1,1593 @@
+    pub fn is_allowed(&self, access: AccessType, user: PublicKey) -> bool {
+        use AccessType::*;
+        use Map::*;
+        // Public flavours automatically allows all reads.
+        match (self, access) {
+            (PublicSentried(_), Read) | (Public(_), Read) => return true,
+            _ => (),
+        }
+        match (self, access) {
+            (PublicSentried(data), Insert)
+            | (PublicSentried(data), Update)
+            | (PublicSentried(data), Delete)
+            | (PublicSentried(data), ModifyPermissions) => data.is_allowed(user, access),
+            (Public(data), Insert)
+            | (Public(data), Update)
+            | (Public(data), Delete)
+            | (Public(data), ModifyPermissions) => data.is_allowed(user, access),
+            (PrivateSentried(data), Insert)
+            | (PrivateSentried(data), Update)
+            | (PrivateSentried(data), Delete)
+            | (PrivateSentried(data), HardUpdate)
+            | (PrivateSentried(data), HardDelete)
+            | (PrivateSentried(data), ModifyPermissions) => data.is_allowed(user, access),
+            (Private(data), Insert)
+            | (Private(data), Update)
+            | (Private(data), Delete)
+            | (Private(data), HardUpdate)
+            | (Private(data), HardDelete)
+            | (Private(data), ModifyPermissions) => data.is_allowed(user, access),
+            (PrivateSentried(data), Read) => data.is_allowed(user, access),
+            (Private(data), Read) => data.is_allowed(user, access),
+            _ => false,
+        }
+    }
+
+    /// Returns the address.
+    pub fn address(&self) -> &Address {
+        use Map::*;
+        match self {
+            PublicSentried(data) => data.address(),
+            Public(data) => data.address(),
+            PrivateSentried(data) => data.address(),
+            Private(data) => data.address(),
+        }
+    }
+
+    /// Returns the kind.
+    pub fn kind(&self) -> Kind {
+        self.address().kind()
+    }
+
+    /// Returns the xor name.
+    pub fn name(&self) -> &XorName {
+        self.address().name()
+    }
+
+    /// Returns the tag type.
+    pub fn tag(&self) -> u64 {
+        self.address().tag()
+    }
+
+    /// Returns true if this instance is public.
+    pub fn is_public(&self) -> bool {
+        self.kind().is_public()
+    }
+
+    /// Returns true if this instance is private.
+    pub fn is_private(&self) -> bool {
+        self.kind().is_private()
+    }
+
+    /// Returns true if this instance employs concurrency control.
+    pub fn is_sentried(&self) -> bool {
+        self.kind().is_sentried()
+    }
+
+    /// Returns true if the provided user (identified by their public key) is the current owner.
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        use Map::*;
+        match self {
+            PublicSentried(data) => data.is_owner(user),
```

Comment:
considering it is repeated again and again, might be worth using a macro like https://github.com/maidsafe/routing/blob/fleming/src/state_machine.rs#L36 (but a bit simpler), it will make all these function trivial and avoid duplication.